From 148f5fb62ba1ab6da66afd8c35049efa3f8225ad Mon Sep 17 00:00:00 2001
From: barracuda156 <vital.had@gmail.com>
Date: Sun, 14 Jan 2024 06:06:39 +0800
Subject: [PATCH 05/10] 
 dom/media/webspeech/synth/cocoa/OSXSpeechSynthesizerService.mm: fix for gcc

---
 .../synth/cocoa/OSXSpeechSynthesizerService.mm      | 13 ++++++++-----
 xpcom/threads/MozPromise.h                          |  6 ++++++
 2 files changed, 14 insertions(+), 5 deletions(-)

diff --git dom/media/webspeech/synth/cocoa/OSXSpeechSynthesizerService.mm dom/media/webspeech/synth/cocoa/OSXSpeechSynthesizerService.mm
index 1c13198563..58c932b6f9 100644
--- dom/media/webspeech/synth/cocoa/OSXSpeechSynthesizerService.mm
+++ dom/media/webspeech/synth/cocoa/OSXSpeechSynthesizerService.mm
@@ -4,6 +4,9 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// Fix glitch in MozPromise.h on .mm files with C++11 lambdas.
+#define GCC_BROKEN_LAMBDAS
+
 #include "nsISupports.h"
 #include "nsServiceManagerUtils.h"
 #include "nsObjCExceptions.h"
@@ -17,7 +20,7 @@
 
 #import <Cocoa/Cocoa.h>
 
-#if defined(MAC_OS_X_VERSION_10_5) && (MAC_OS_X_VERSION_MAX_ALLOWED <= MAC_OS_X_VERSION_10_5)
+#if defined(MAC_OS_X_VERSION_10_5) && (MAC_OS_X_VERSION_MAX_ALLOWED <= MAC_OS_X_VERSION_10_6)
 @protocol NSSpeechSynthesizerDelegate
 @end
 #endif
@@ -285,7 +288,9 @@ EnumVoicesRunnable::Run()
   NSArray* voices = [NSSpeechSynthesizer availableVoices];
   NSString* defaultVoice = [NSSpeechSynthesizer defaultVoice];
 
-  for (NSString* voice in voices) {
+  int i;
+  for (i=0; i<[voices count]; i++) {
+    NSString *voice = [voices objectAtIndex:i];
     OSXVoice item;
 
     NSDictionary* attr = [NSSpeechSynthesizer attributesForVoice:voice];
@@ -296,9 +301,7 @@ EnumVoicesRunnable::Run()
 
     nsCocoaUtils::GetStringForNSString([attr objectForKey:NSVoiceName], item.mName);
 
-    nsCocoaUtils::GetStringForNSString(
-      [attr objectForKey:NSVoiceLocaleIdentifier], item.mLocale);
-    item.mLocale.ReplaceChar('_', '-');
+    item.mLocale.AssignLiteral("en-us");
 
     item.mUri.AssignLiteral("urn:moz-tts:osx:");
     item.mUri.Append(identifier);
diff --git xpcom/threads/MozPromise.h xpcom/threads/MozPromise.h
index dab350f422..f0fe4c168c 100644
--- xpcom/threads/MozPromise.h
+++ xpcom/threads/MozPromise.h
@@ -255,6 +255,9 @@ public:
 
   static RefPtr<AllPromiseType> All(AbstractThread* aProcessingThread, nsTArray<RefPtr<MozPromise>>& aPromises)
   {
+// Work around older gcc 4.8 versions misinterpreting C++11 lambdas in Obj-C++
+// files as Objective-C messages. This only works if we don't call "All".
+#ifndef GCC_BROKEN_LAMBDAS
     RefPtr<AllPromiseHolder> holder = new AllPromiseHolder(aPromises.Length());
     for (size_t i = 0; i < aPromises.Length(); ++i) {
       aPromises[i]->Then(aProcessingThread, __func__,
@@ -263,6 +266,9 @@ public:
       );
     }
     return holder->Promise();
+#else
+    MOZ_CRASH("objective-c++ clash with MozPromise:All");
+#endif
   }
 
   class Request : public MozPromiseRefcountable
-- 
2.43.0

